### PR TITLE
Visitor can view a collection of articles

### PR DESCRIPTION
**Pull request by johnnysee**
Link to PT: https://www.pivotaltracker.com/story/show/179886990

## User Story

```
As a visitor,
In order to get an overview of the website's news
I would like to be able to see a collection of article previews
```

## Tasks

- Create a display News component
- `GET` request to API with `useEffect` hook
- Install Axios
- Add fixture file for stubbing
- Add a `title`, `lede`, `authors`(an array for multiple authors), `created_at` & `updated_at` 
- `.map` used on fixture file json to list articles
- error handling function. add `try` `catch` to the axios call requests
## Learning subjects
- async and await trouble shooting when only returning promise
- Not to feature creep just because it will solve the issues in the moment
- Repetition of map functions is a good idea
- The value of well discussed PT instructions in scope, minimizes risk of future mistakes 

Skärmavbild 2021-10-15 kl  00 04 32

### CODE REVIEW

File: src/modules/apiHelper.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import axios from "axios";
+
+const articlesIndex = async () => {
+  try {
+    const response = await axios.get("/api/articles/");
+    return response.data.articles;
+  } catch (error) {
+    errorHandler(error);
+  }
+};
```
Comment (tochman) [inline src/modules/apiHelper.js:12]: This is totally okay, but it is not what I suggested during our discussion about how to extract this functionality. If it was hard to follow my reasoning, you should ask for clarification.

File: src/components/ArticleItem.jsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import { Segment, Grid } from "semantic-ui-react";
+
+const ArticleItem = ({ articleItem, index }) => {
+  return (
+    <Grid.Column width="4" data-cy="news-section">
+      <Segment data-cy={`article-${index}`}>
```
Comment (tochman) [inline src/components/ArticleItem.jsx]: why not use the `id` of the article?

File: src/components/HomePage.jsx [hunk 3]
```diff
@@ -1,11 +1,35 @@
...
+      const articlesResponse = await articlesIndex();
+      setArticles(articlesResponse);
+    })();
+  }, []);
+
+  let articleList = articles.map((articleItem, index) => {
+    return (
+      <ArticleItem
+        articleItem={articleItem}
+        index={index}
```
Comment (tochman) [inline src/components/HomePage.jsx]: Why do you need to pass in this value as a propp?

File: src/components/HomePage.jsx [hunk 4]
```diff
@@ -1,11 +1,35 @@
...
+import ArticleItem from "./ArticleItem";
 
 const HomePage = () => {
+  const [articles, setArticles] = useState([]);
+
+  useEffect(() => {
+    (async () => {
+      const articlesResponse = await articlesIndex();
+      setArticles(articlesResponse);
+    })();
```
Comment (tochman) [inline src/components/HomePage.jsx:13]: This looks fine to me but I would like you to check if React complains by issuing a warning. Check the console output in Cypress.

File: cypress/fixtures/indexRespondsFromApi.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
+  "articles": [
+    {
+      "id": 1,
+      "title": "TKTK",
+      "lede": "TKTKT",
+      "authors": ["bob journalist", "bobette journalist"],
+      "created_at": "20211014-1313",
```
Comment (maxarvid) [inline cypress/fixtures/indexRespondsFromApi.json]: Consider converting this date format to: `"2021-10-04T13:33:56.923Z"`

File: src/components/HomePage.jsx (hunk 4 above)
Comment (dorianbuck) [reply src/components/HomePage.jsx:13]: There is no warning

File: src/components/ArticleItem.jsx (hunk 2 above)
Comment (dorianbuck) [reply src/components/ArticleItem.jsx]: I changed it to id instead of index

File: src/components/HomePage.jsx (hunk 3 above)
Comment (dorianbuck) [reply src/components/HomePage.jsx]: This has been changed!

**Comment by dorianbuck** [comment]
I have implemented changes after these comments. Thank you!

File: cypress/integration/userCanViewACollectionOfArticles.feature.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      statusCode: 200,
+    }).as("indexApiGetRequest");
+
+    cy.visit("/");
+    cy.get("[data-cy=news-section]").as("newsSection");
+  });
+
+
+  it("is expected to display a collection of articles", () => {
+    cy.get("@newsSection").children().should("has.length", 3);
```
Comment (offcast-dev) [inline cypress/integration/userCanViewACollectionOfArticles.feature.js]: "has.length" does that work? it is spelled "have.length" usually.

File: cypress/integration/userCanViewACollectionOfArticles.feature.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  });
+
+  it("is expected to display a list of articles sorted by id/index", () => {
+    cy.get("@newsSection").within(() => {
+      cy.get("[data-cy=article-1]");
+      cy.get("[data-cy=title]").should("be.visible");
+      cy.get("[data-cy=lede]").should("be.visible");
+      cy.get("[data-cy=authors]").should("be.visible");
+      cy.get("[data-cy=created_at]").should("be.visible");
+      cy.get("[data-cy=updated_at]").should("be.visible");
```
Comment (offcast-dev) [inline cypress/integration/userCanViewACollectionOfArticles.feature.js]: not the best assertion as it does not check content. If at some point the bug will appear that displays wrong information in one of those fields, you wont catch it with this tests.

File: cypress/integration/userCanViewACollectionOfArticles.feature.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    });
+  });
+  it("is expected to return a http status response", () => {
+    cy.wait("@indexApiGetRequest").its("response.statusCode").should("eq", 200);
+  });
```
Comment (offcast-dev) [inline cypress/integration/userCanViewACollectionOfArticles.feature.js:32]: cool

File: src/modules/apiHelper.js (hunk 1 above)
Comment (offcast-dev) [reply src/modules/apiHelper.js:12]: JavaScript allows to store functions inside an object. What you can do here to improve structure of a module is to create `Articles` object that holds your `index` function and `export default Articles`.

That means when you want to access `index` function in your component you import `Articles` and access it like `Articles.index`

**Review by offcast-dev — COMMENTED** [review]
Some changes

File: cypress/integration/userCanViewACollectionOfArticles.feature.js (hunk 6 above)
Comment (dorianbuck) [reply cypress/integration/userCanViewACollectionOfArticles.feature.js]: It does work with has as well. For best practice I will change this to have instead

File: cypress/integration/userCanViewACollectionOfArticles.feature.js (hunk 7 above)
Comment (dorianbuck) [reply cypress/integration/userCanViewACollectionOfArticles.feature.js]: These test have been altered to check for the actual content.

File: src/modules/apiHelper.js (hunk 1 above)
Comment (dorianbuck) [reply src/modules/apiHelper.js:12]: This has now been changed. Thank you!

**Review by offcast-dev — APPROVED** [review]
Good job!
